### ISSUE TITLE
Exclude option ignored when directly calling dir/file

### ISSUE THREAD

**Issue by jhossbach** [charliermarsh/ruff#1339]
MWE:
```toml
[tool.ruff]
exclude = [ "blah" ]
```
`blah/test.py`:
```python
import os
```
Running ruff on `.` returns nothing, but running `ruff ./blah` or `ruff ./blah/test.py` will return the diagnostics (F401) even if blah was excluded. Is this intentional?

**Comment by jhossbach** [comment]
*To be more precise: I think this is unwanted because it will mess with parsing the file via stdin and then giving the `stdin-filename`

**Comment by charliermarsh** [comment]
Yeah it's intentional. There's a `--force-exclude` flag you can provide on the command-line or in `pyproject.toml` to disable this.

**Comment by charliermarsh** [comment]
This is how Black behaves and I think it's pretty reasonable -- if you pass a file directly, we should lint it even if it's marked as an exclusion, since in most cases that's intentional behavior.

**Comment by charliermarsh** [comment]
But in some settings, it makes sense to turn that off (for example, with `pre-commit`).

**Comment by charliermarsh** [comment]
(And, similarly, it might make sense to always pass that flag in the LSP context.)

**Comment by jhossbach** [comment]
In that case I will keep this behavior in `python-lsp-ruff` rather than unintuitively changing it

**Comment by charliermarsh** [comment]
I'll probably change `ruff-lsp` to always pass that flag, it seems confusing in the context of an IDE.

**Comment by jhossbach** [comment]
I guess you're right, it would also require to add that option to the LSP configuration which should be redundant.

**Comment by charliermarsh** [comment]
Need to try out the `ruff --force-exclude subdir/file1.py` case to understand what’s going on there, but I think you’re probably right that we’re not respecting exclusions at all when running via stdin (which I misstated above). We need to make code changes to support that, most likely.

**Comment by charliermarsh** [comment]
(And no need to apologize, sorry for misleading you above regarding stdin!)

**Comment by charliermarsh** [comment]
The `ruff --force-exclude subdir/file1.py` issue is just sort of a quirk of how the pattern-matching works. If you provide `subdir` as the exclude, then we exclude any file or directory that ends in `subdir`. If you provide `subdir` as the folder to check on the command-line, then we exclude it due to that match and avoid recursing into it. If you provide `subdir/file1.py`, we never actually check `subdir`, just `file1.py` and `subdir/file1.py`, which is why `subdir/*` actually _does_ work.

I think this can be improved, but need to consider what the right fix is...

**Comment by jhossbach** [comment]
How about a gitignore-like style? It would mean that `foo/` would exclude a dir and everything beneath, while `foo/*` would match everything directly beneath (so `foo/bar/test.py`) would not be excluded (of course we can relax this part).
`subdir/file1.py` would match this file and this file only, while `file1.py` would match `file1.py` only in the first directory, while `**file1.py` matches `file1.py` in every directory.

**Comment by jhossbach** [comment]
Seems to work :+1: I'll open up a new issue for the extension to when passing the file through stdin.
